skfp/fingerprints/lingo.py

```diff
@@ -0,0 +1,183 @@
+        "substring_length": [Interval(Integral, 1, None, closed="left")],
+        "count": ["boolean"],
+    }
+
+    def __init__(
+        self,
+        fp_size: int = 1024,
+        substring_length: int = 4,
+        count: bool = False,
+        sparse: bool = False,
+        n_jobs: Optional[int] = None,
+        batch_size: Optional[int] = None,
+        verbose: int = 0,
+    ):
+        super().__init__(
+            n_features_out=fp_size,
+            sparse=sparse,
+            n_jobs=n_jobs,
+            batch_size=batch_size,
+            verbose=verbose,
+        )
+        self.fp_size = fp_size
+        self.substring_length = substring_length
+        self.count = count
+
+    def transform(
+        self, X: Sequence[Union[str, Mol]], copy: bool = False
+    ) -> Union[np.ndarray, csr_array]:
+        """
+        Compute Lingo fingerprints.
+
+        Parameters
+        ----------
+        X : {sequence, array-like} of shape (n_samples,)
+            Sequence containing SMILES strings or RDKit Mol objects.
+
+        copy : bool, default=False
+            Whether to copy X or modify it in place.
+
+        Returns
+        -------
+        X: {ndarray, sparse matrix} of shape (n_samples, self.fp_size)
+            2D array containing Lingo fingerprints.
+        """
+        return super().transform(X, copy)
+
+    def smiles_to_dicts(
+        self, X: Sequence[Union[str, Mol]]
+    ) -> list[defaultdict[str, int]]:
+        """
+        Convert SMILES strings to dictionaries of substring counts.
+
+        Parameters
+        ----------
+        X: {sequence, array-like} of shape (n_samples,)
+            Sequence containing SMILES strings or RDKit Mol objects.
+
+        Returns
+        -------
+        result: list[defaultdict[str, int]]
+            List of dictionaries containing substring counts.
+        """
+        X = ensure_smiles(X)
+        X = [re.sub(r"[123456789]", "0", x) for x in X]
+        X = [re.sub(r"Cl", "L", x) for x in X]
+        X = [re.sub(r"Br", "R", x) for x in X]
+
+        result = []
+        for smiles in X:
+            result_dict: defaultdict[str, int] = defaultdict(int)
+            for i in range(len(smiles) - self.substring_length + 1):
+                result_dict[smiles[i : i + self.substring_length]] += 1
+            result.append(result_dict)
+        return result
+
+    def _calculate_fingerprint(
+        self, X: Sequence[Union[str, Mol]]
+    ) -> Union[np.ndarray, csr_array]:
+        X = self.smiles_to_dicts(X)
+        X = self._dicts_to_array(X)
+        return csr_array(X) if self.sparse else X
+
+    def _dicts_to_array(self, X: list[defaultdict[str, int]]) -> np.ndarray:
+        dtype = np.uint32 if self.count else np.uint8
+        result = np.zeros((len(X), self.fp_size), dtype=dtype)
+        for i, dictionary in enumerate(X):
+            for key, value in dictionary.items():
+                string_bytes = key.encode()
```

Comment:
Specify "utf-8" explicitly. Doesn't `bandit` pre-commit hook complain about this?

Reply:
python3 default encoding is UTF* as in [docs](https://docs.python.org/3/howto/unicode.html#the-string-type). Of course I can write it explicitly. Tests have passed, so no - bandit didn't complain.

Reply:
Ok